### ISSUE TITLE
Make list-datum fast for consumption via the UI

### ISSUE THREAD

**Issue by sjezewski** [pachyderm/pachyderm#2151]
Whether that be straight up pagination, streaming, etc.

First some benchmarks.

On the UI capturing the JS / render times:

```
100datums

Total: 1.95 s
1571.1 msScripting
46.5 msRendering
12.1 msPainting
25.1 msOther
298.3 msIdle


500datums

Total: 7.06 s
5375.7 msScripting
189.9 msRendering
4.1 msPainting
24.5 msOther
1467.8 msIdle

5k datums 

cant get profile ... just crashes
```

For pachd:

```
100 datums

starting listFilegetTreeForFile took 1.288629ms for file commit:<repo:<name:"pipeline147989210ff7" > id:"411c5e3476cd44a5b0b7038c1922ad7b" > path:"/" 
tree.list, took 86.371µs
finished listFile, took 1.787325ms
time="2017-08-10T21:56:08Z" level=info msg="Response contains 101 objects; logging the first 10" 
listdatum - time to do listfile 2.273537ms
listdatum - time to do all diff-files 123.694274ms
listdatum - time to sort 93.623µs
2017-08-10T21:56:08Z INFO pps.API.ListDatum {"duration":0.126644456,"error":null,"...fs"}}]}} 

500 datums

starting listFilegetTreeForFile took 1.230517ms for file commit:<repo:<name:"pipelinedd5e9a4b9849" > id:"9ddd94576ef0489d851cd2a8dcacf8dd" > path:"/" 
tree.list, took 328.578µs
finished listFile, took 2.318503ms
time="2017-08-10T21:57:48Z" level=info msg="Response contains 501 objects; logging the first 10" 
listdatum - time to do listfile 3.412762ms
listdatum - time to do all diff-files 655.492445ms
listdatum - time to sort 568.3µs
2017-08-10T21:57:49Z INFO pps.API.ListDatum {"duration":0.660000095,"fs"}}]}} 

5k datums

starting listFilegetTreeForFile took 3.135169ms for file commit:<repo:<name:"pipelinea897209e2edd" > id:"85c53c91105d4f22b88675a8a09d2d79" > path:"/" 
tree.list, took 2.81488ms
time="2017-08-10T21:58:44Z" level=info msg="Response contains 5001 objects; logging the first 10" 
finished listFile, took 9.019324ms
listdatum - time to do listfile 14.432963ms
listdatum - time to do all diff-files 5.665259848s
listdatum - time to sort 15.084715ms
2017-08-10T21:58:50Z INFO pps.API.ListDatum {"duration":5.695363723,...}
```

So the pachd response is fast enough for ~100 datums.

There seems to be a large overhead on the UI side in JS parsing / constructing the UI (rendering times are ok).

**Comment by sjezewski** [comment]
Additionally profiling the mem used on the UI side:

```
100 datums: + 28MB
500 datums: + 55MB
5k datums: + 590MB
```

So I have some concerns about a streaming solution. Such a solution would provide the UI the data sooner ... but the latency from the pachd call isn't really the issue ... most of the time spent here is in JS land (parsing / prepping). So while a streaming call would provide the data sooner, we'd have to make it so that the JS caches the data that was not displayed. If we're designing this to work for on the order of 10k datums, that means we'll have ~1GB of mem used by the UI. That seems like too much.

---

So I see two other options.

A) Paginate

This would require making the `ListFile` api call truly paginate. Basically you'd provide an offset, and it would return a batch of files w that offset from the total count.

This would require some tricky stuff on the backend. Perhaps the 'heavy write' that @msteffen discussed yesterday (when writing a hashtree ... also write sub-hashtrees of standard chunk sizes so that we can load those directly from obj store ... and perhaps we only do this for stats hashtrees not normal files).

But this seems tricky.

B) Make `list-datum` fast enough it just doesn't matter

Right now the longest part of `list-datum` is the part where we diff the files / gather each datum. This accounts for 90% of the time spent in `list-datum`. If we could optimize this to be an order of magnitude or two faster, then we could do dumb pagination (generate all the results ... only return the slice that matters).

This has the advantage that we want to fix this codepath anyway. @jdoliner pointed out a correctness bug with this part of the code (where we incorrectly mark a datum as new when it is in fact skipped). The solution would be instead of doing `diff-file` using an `inspect-file` to determine datum status. If the hashtree for this commit is in the cache (which it would be after checking the first datum), we would avoid any trips to fetch objects from obj storage. (I think `diff-file` gets 2 hashtrees (one for parent commit, one for current commit) and doesn't fetch any objs from obj storage ... so this may not be a huge perf win).

But, since we want to make this change anyway, I think it makes sense to start there first, and also see if we can squeeze any performance juice out there as well.

**Comment by sjezewski** [comment]
Offline discussion

---

1) We're going to proceed first w the correctness fix

This will write a `skipped` file to the stats tree if the datum was skipped. This will be done on the worker master

2) We can optimize the code in `list-datum` 

After we read all the files (to get all the datums), we can already get the state (bcz each fileinfo is a dir w a children field ... we can already test for the existence of the failed/skipped files) and sort before doing the expensive operation of getting each datum

**Comment by sjezewski** [comment]
This is fixed in #2171
